search.go

```diff
@@ -231,6 +231,20 @@ type SearchRequest struct {
 	Controls     []Control
 }
 
+func (req *SearchRequest) appendBaseDN(dn string) appendDnRequest {
+	return &SearchRequest{
```

Comment:
This is an unnecessary allocation. You're working with a pointer anyway, why not edit object in place and simply change the field? E.g.
````go
func (req *SearchRequest) appendBaseDN(dn string) {
    req.BaseDN = appendDN(req.BaseDN, dn)
}
````

This also woldn't require us to constantly update this method in case new fields are added/changed.

Reply:
As a user, I don't expect my input to be modified by the library, at least, the original version won't. However, there is a way to implement a clone and edit method

Reply:
Most of the operations in this library mutate the given input.  Let's stick with the same approach for now, and if we decide to clone and edit user input in the future we can do that everywhere, rather than inconsistently.